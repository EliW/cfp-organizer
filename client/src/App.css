@charset "UTF-8";
@import url(https://fonts.googleapis.com/css?family=Open+Sans:300,400,700);
body {
  font-family: 'Open Sans', sans-serif;
  font-weight: 300;
  line-height: 1.42em;
  color:#A7A1AE;
  background-color:#1F2739;
}

a {
  color: #FB667A;
}

.Table {
  display: flex;
  flex-direction: column;
  overflow: hidden;
  width: 80%;
  margin: 0 auto;
}

.Table-header {
  display: flex;
  color: #fff;
}

.Table-headerCell {
  flex: 1;
  cursor: pointer;
  background-color: #1F2739;
  font-weight: bold;
  font-size: 1em;
  text-align: left;
  color: #5496b3;
}

.EventRow {
  display: flex;
}

.EventRow-cell {
  flex: 1;
  font-weight: normal;
  font-size: 1em;
  -webkit-box-shadow: 0 2px 2px -2px #0E1119;
  -moz-box-shadow: 0 2px 2px -2px #0E1119;
  box-shadow: 0 2px 2px -2px #0E1119;
  border: 1px solid #1F2739;
  border-radius: 2px;
}

.EventRow .EventRow-cell {
  background-color: #323C50;
}

.EventRow:nth-child(even) .EventRow-cell {
  background-color: #2C3446;
}

.Table-headerCell, .EventRow-cell {
  padding: 1em 2em;
}