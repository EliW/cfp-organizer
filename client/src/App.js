import React, { Component } from 'react';
import moment from 'moment';
import { Loading } from './components/Loading.js';
import { Table } from './components/Table.js';
import { EventRow } from './components/EventRow.js';
import { MobileEventRow } from './components/MobileEventRow.js';
import { Map } from './components/Map.js';
import { EventPopup } from './components/EventPopup.js';
import './App.css';

class App extends Component {
  state = {
    data: null, 
    activeView: 'list',
    loading: true,
    tableHeaders: [
      {
        name: 'name',
        label: 'Name',
        sortable: true,
        filterable: true,
      },
      {
        name: 'location',
        label: 'Location',
        sortable: true,
        filterable: true,
        filterKeys: ['location', 'continent', 'country', 'countryCode', 'city']
      },
      {
        name: 'date',
        label: 'Date',
        sortable: true,
        sortKey: 'momentDate',
        filterable: false,
      },
      {
        name: 'cfpClose',
        label: 'CFP Close Date',
        sortable: true,
        sortKey: 'momentCfpClose',
        filterable: false,
      },
      {
        name: 'eventTags',
        label: 'Event Tags',
        sortable: false,
        filterable: true,
      }
    ],
    mobileHeaders: [
      {
        name: 'events',
        label: 'Events',
        sortable: true,
        sortKey: 'name',
        filterable: true,
        filterKeys: ['name', 'location', 'continent', 'country', 'countryCode', 'city', 'eventTags']
      }
    ]
  }
  
  componentDidMount() {
    this.callApi()
      .then((response) => {
        // data cleanup
        return response.events.map((e, i) => {
          return {
            ...e,
            momentDate: !!e.date ? moment(e.date, 'MMMM DD, YYYY') : null,
            momentCfpClose: !!e.cfpClose ? moment(e.cfpClose, 'MMMM DD, YYYY HH: mm UTC') : null,
            key: i
          }
        })
      })
      .then(events => this.setState({ data: events, loading: false }))
      .catch(err => console.log(err));
  }

  callApi = async () => {
    const response = await fetch('/api/openCfps');
    const body = await response.json();

    if (response.status !== 200) throw Error(body.message);

    return body;
  };

  setActiveView = (view) => {
    this.setState({ activeView: view })
  }

  setActiveDataItem = (dataItemKey) => {
    if (this.state.activeDataItem === dataItemKey) {
      this.setState({ activeDataItem: null })
    } else {
      this.setState({ activeDataItem: dataItemKey })
    }
  }

  render() {
    return (
      <div className="App">
        <div className="App-header">
          <h1>CFP Organizer</h1>
          <h2>Sort and filter Papercall CFPs</h2>
          <ul className='ViewList'>
            <li className={`ViewList-item ${this.state.activeView === 'list' ? 'ViewList-item--active' : ''}`} onClick={() => this.setActiveView('list')}>List</li>
            <li className={`ViewList-item ${this.state.activeView === 'map' ? 'ViewList-item--active' : ''}`} onClick={() => this.setActiveView('map')}>Map</li>
          </ul>
        </div>
        { this.state.loading &&
          <Loading />
        }
        { !this.state.loading && this.state.activeView === 'list' &&
          <React.Fragment>
            <Table
              className='Table--mobile'
              headers={this.state.mobileHeaders}
              data={this.state.data}
              rowComponent={(dataItem, i) => (
                <MobileEventRow 
                  className='' 
                  cellClassName='MobileCell' 
                  key={dataItem.key} 
                  event={dataItem} 
                  isActive={this.state.activeDataItem === dataItem.key}
                  onClick={() => this.setActiveDataItem(dataItem.key)} 
                />
              )}
            />
  
            <Table
              headers={this.state.tableHeaders}
              data={this.state.data}
              rowComponent={(dataItem, i) => <EventRow className='Table-row' cellClassName='Table-cell' key={dataItem.key} event={dataItem} />}
            />
          </React.Fragment>
        }
        { !this.state.loading && this.state.activeView === 'map' &&
          <Map
            data={this.state.data}
            popupComponent={(dataItem) => <EventPopup event={dataItem} />}
          />
        }
      </div>
    );
  }
}

export default App;