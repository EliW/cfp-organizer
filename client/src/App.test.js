import React from 'react';
import { shallow } from 'enzyme'
import App from './App';

let subject

function loadSubject(props = {}) {
  subject = shallow(<App />);
  subject.instance().callApi = jest.fn({ events: [] });
  subject.update();
}

it('renders', () => {
  let subject = shallow(<App />);
  subject.instance().callApi = jest.fn({events: []});
  subject.update();
  expect(subject).toMatchSnapshot()
})

describe('setActiveView', () => {
  it('sets the state', () => {
    loadSubject()
    subject.instance().setActiveView('map')
    expect(subject.state('activeView')).toEqual('map')
  })
})

describe('setActiveDataItem', () => {
  describe('when the dataItemKey does not match state', () => {
    it('sets the state', () => {
      loadSubject()
      subject.instance().setActiveDataItem(0)
      expect(subject.state('activeDataItem')).toEqual(0)
    })
  })

  describe('when the dataItemKey matches state', () => {
    it('sets the state to null', () => {
      loadSubject()
      subject.instance().setState({ activeDataItem: 0 })
      subject.update()
      subject.instance().setActiveDataItem(0)
      expect(subject.state('activeDataItem')).toBeNull()
    })
  })
})