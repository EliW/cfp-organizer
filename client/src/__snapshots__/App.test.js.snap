// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="App-header"
>
          <h1>
                    CFP Organizer
          </h1>
          <h2>
                    Sort and filter open CFPs
          </h2>
          <ul
                    className="ViewList"
          >
                    <li
                              className="ViewList-item ViewList-item--active"
                              onClick={[Function]}
                    >
                              List
                    </li>
                    <li
                              className="ViewList-item "
                              onClick={[Function]}
                    >
                              Map
                    </li>
          </ul>
</div>,
        <Unknown />,
        false,
        false,
        <p
          className="Created u-small"
>
          Created by Mercedes Bernard |
           
          <a
                    href="https://github.com/mercedesb"
          >
                    Github
          </a>
           |
           
          <a
                    href="https://twitter.com/mercedescodes"
          >
                    Twitter
          </a>
</p>,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h1>
              CFP Organizer
</h1>,
            <h2>
              Sort and filter open CFPs
</h2>,
            <ul
              className="ViewList"
>
              <li
                            className="ViewList-item ViewList-item--active"
                            onClick={[Function]}
              >
                            List
              </li>
              <li
                            className="ViewList-item "
                            onClick={[Function]}
              >
                            Map
              </li>
</ul>,
          ],
          "className": "App-header",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "CFP Organizer",
            },
            "ref": null,
            "rendered": "CFP Organizer",
            "type": "h1",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Sort and filter open CFPs",
            },
            "ref": null,
            "rendered": "Sort and filter open CFPs",
            "type": "h2",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <li
                  className="ViewList-item ViewList-item--active"
                  onClick={[Function]}
>
                  List
</li>,
                <li
                  className="ViewList-item "
                  onClick={[Function]}
>
                  Map
</li>,
              ],
              "className": "ViewList",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "List",
                  "className": "ViewList-item ViewList-item--active",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "List",
                "type": "li",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Map",
                  "className": "ViewList-item ",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "Map",
                "type": "li",
              },
            ],
            "type": "ul",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      false,
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Created by Mercedes Bernard |",
            " ",
            <a
              href="https://github.com/mercedesb"
>
              Github
</a>,
            " |",
            " ",
            <a
              href="https://twitter.com/mercedescodes"
>
              Twitter
</a>,
          ],
          "className": "Created u-small",
        },
        "ref": null,
        "rendered": Array [
          "Created by Mercedes Bernard |",
          " ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Github",
              "href": "https://github.com/mercedesb",
            },
            "ref": null,
            "rendered": "Github",
            "type": "a",
          },
          " |",
          " ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Twitter",
              "href": "https://twitter.com/mercedescodes",
            },
            "ref": null,
            "rendered": "Twitter",
            "type": "a",
          },
        ],
        "type": "p",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="App-header"
>
            <h1>
                        CFP Organizer
            </h1>
            <h2>
                        Sort and filter open CFPs
            </h2>
            <ul
                        className="ViewList"
            >
                        <li
                                    className="ViewList-item ViewList-item--active"
                                    onClick={[Function]}
                        >
                                    List
                        </li>
                        <li
                                    className="ViewList-item "
                                    onClick={[Function]}
                        >
                                    Map
                        </li>
            </ul>
</div>,
          <Unknown />,
          false,
          false,
          <p
            className="Created u-small"
>
            Created by Mercedes Bernard |
             
            <a
                        href="https://github.com/mercedesb"
            >
                        Github
            </a>
             |
             
            <a
                        href="https://twitter.com/mercedescodes"
            >
                        Twitter
            </a>
</p>,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h1>
                CFP Organizer
</h1>,
              <h2>
                Sort and filter open CFPs
</h2>,
              <ul
                className="ViewList"
>
                <li
                                className="ViewList-item ViewList-item--active"
                                onClick={[Function]}
                >
                                List
                </li>
                <li
                                className="ViewList-item "
                                onClick={[Function]}
                >
                                Map
                </li>
</ul>,
            ],
            "className": "App-header",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "CFP Organizer",
              },
              "ref": null,
              "rendered": "CFP Organizer",
              "type": "h1",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Sort and filter open CFPs",
              },
              "ref": null,
              "rendered": "Sort and filter open CFPs",
              "type": "h2",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <li
                    className="ViewList-item ViewList-item--active"
                    onClick={[Function]}
>
                    List
</li>,
                  <li
                    className="ViewList-item "
                    onClick={[Function]}
>
                    Map
</li>,
                ],
                "className": "ViewList",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "List",
                    "className": "ViewList-item ViewList-item--active",
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": "List",
                  "type": "li",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Map",
                    "className": "ViewList-item ",
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": "Map",
                  "type": "li",
                },
              ],
              "type": "ul",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Created by Mercedes Bernard |",
              " ",
              <a
                href="https://github.com/mercedesb"
>
                Github
</a>,
              " |",
              " ",
              <a
                href="https://twitter.com/mercedescodes"
>
                Twitter
</a>,
            ],
            "className": "Created u-small",
          },
          "ref": null,
          "rendered": Array [
            "Created by Mercedes Bernard |",
            " ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Github",
                "href": "https://github.com/mercedesb",
              },
              "ref": null,
              "rendered": "Github",
              "type": "a",
            },
            " |",
            " ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Twitter",
                "href": "https://twitter.com/mercedescodes",
              },
              "ref": null,
              "rendered": "Twitter",
              "type": "a",
            },
          ],
          "type": "p",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
